docs/auth/oidc.md

```diff
@@ -0,0 +1,231 @@
+instance of something, in our case it will be a function that requests some backstage
+APIs and use them to create an instance of an OIDC provider.
+
+Please note that this function only runs (creates the instance) when somewhere else in
+the app you request the DI to give you an instance of the OIDC provider using the API ref
+defined above, and the DI will only run this function the first time, from then on any
+other DI injection will just receive the same instance created the first time, basically
+the instance is cached by the DI library, a singleton.
+
+Let's add our OIDC factory to the APIs array in the `packages/app/src/apis.ts` file:
+
+```diff
++ import { OAuth2 } from '@backstage/core-app-api';
+
+export const apis: AnyApiFactory[] = [
++   createApiFactory({
++    api: azureOIDCAuthApiRef,
++    deps: {
++      discoveryApi: discoveryApiRef,
++      oauthRequestApi: oauthRequestApiRef,
++      configApi: configApiRef,
++    },
++    factory: ({ discoveryApi, oauthRequestApi, configApi }) =>
++      OAuth2.create({
++        discoveryApi,
++        oauthRequestApi,
++        provider: {
++          id: 'oidc', // This has to be 'oidc' or OAuth2 will not use oidc protocol
++          title: 'OIDC provider',
++          icon: () => null,
++        },
++        environment: configApi.getOptionalString('auth.environment'),
++        defaultScopes: [
++          'openid',
++          'profile',
++          'email',
++        ],
++      }),
++  }),
+
+```
+
+Please note we're importing the `OAuth2` class from `@backstage/core-app-api` effectively
+delegating the authentication to it (yes it can handle OIDC as well). Also we're using
+the `oidc` ID to tell `OAuth2` to use the OIDC protocol, and added the default scopes to
+request ID, profile, email and user read permissions.
+
+### The Resolver
+
+Resolvers exist to map user identity from the 3rd party (in this case an azure IDP
+provider) to the backstage user identity, for a detailed explanation check the
+[Identity Resolver][1] page, it explains how to write a custom resolver as well as
+linking the built in resolvers of backstage.
+
+As an example if you're setting up OIDC provider with Azure IDP, you could reuse
+the built in resolvers, or create one yourself in `packages/backend/src/plugins/auth.ts`:
+
+```diff
+import {
+  DEFAULT_NAMESPACE,
++ stringifyEntityRef,
+} from '@backstage/catalog-model';
+
+export default async function createPlugin(
+  env: PluginEnvironment,
+): Promise<Router> {
+  return await createRouter({
+    logger: env.logger,
+    config: env.config,
+    database: env.database,
+    discovery: env.discovery,
+    tokenManager: env.tokenManager,
+    providerFactories: {
+      ...defaultAuthProviderFactories,
++     azureOIDC: providers.oidc.create({
++       signIn: {
++         resolver(info, ctx) {
++           const userRef = stringifyEntityRef({
++             kind: 'User',
++             name: info.profile.email!,
```

Comment:
Let's not use the email as an example, it's a poor user entity name as it's not supported by default. Perhaps the `sub` from `userinfo` instead?

Reply:
I'm using the same name as `commonByEmailLocalPartResolver` or `commonByEmailResolver`, both grab the email, the first one splits by `@` and grabs the first part, while the other uses it directly, so I thought I would keep the same pattern, should I change regardless?

Reply:
I looked to the other providers, both `emailMatchingUserEntityAnnotation` versions use `info.profile.email` as well so...

Reply:
We could do it by email local part, but feeling that's a bunch of extra code that perhaps distracts the example. Although looking things up in the catalog based on the email could be pretty sweet. Main thing I wanna avoid is email as entity name

Reply:
So instead of a custom one should I just use `commonByEmailLocalPartResolver`?

Reply:
I think I avoided it because it would require a catalog user to match properly, that's why I used the custom resolver, want me to add the catalog user part?